### PR TITLE
DynamoDb Adapter Api discussion continued

### PR DESCRIPTION
**Pull request by Ankcorn**
This is hopefully a more productive forum to discuss https://github.com/laconiajs/laconia/issues/152

This PR is open to examine changes to the dynamodb adapter so that it can serve a wider veriety of use cases.

### CODE REVIEW

**Comment by Ankcorn** [comment]
Proposal 1:

Have the event argument of the lambda be 

```
[{
      eventName: 'Modify',
      StreamViewType: 'NEW_IMAGE',
      Keys: {},
      NewImage:  { id: '12rt-34g3g', name: 'cat' },
      OldImage: {}
}]
```

Pro's
* Very Simple Integration
* Supports most dynamodb stream use cases

Con's
* Some aws interface details leaking through

File: packages/laconia-event/src/DynamoDbStreamRecord.js [hunk 1]
```diff
@@ -1,18 +1,36 @@
-const AWS = require("aws-sdk");
+require("aws-sdk/clients/dynamodb");
+const converter = require("aws-sdk/lib/dynamodb/converter");
 
 module.exports = class DynamoDbStreamRecord {
   constructor(data) {
     this.data = data;
```
Comment (hugosenari) [inline packages/laconia-event/src/DynamoDbStreamRecord.js:6]: Maybe you could use destructuring here:
```javascript
this.data = data || {};
```
I did some test here and if you call `converter.unmarshall(undefined)` (same for null and {}) it returns `{}`

Your result will look like this:
```javascript
return {
  eventName: this.eventName,
  StreamViewType: data.StreamViewType,
  Keys: converter.unmarshall(data.Keys),
  NewImage: converter.unmarshall(data.NewImage),
  OldImage: converter.unmarshall(data.OldImage)
};
```

File: packages/laconia-event/src/DynamoDbStreamRecord.js [hunk 2]
```diff
@@ -1,18 +1,36 @@
...
+    return converter.unmarshall(this.newImage);
   }
 
   get newImage() {
     return this.data.NewImage;
   }
 
+  get parseRecord() {
+    return {
+      eventName: this.eventName,
```
Comment (hugosenari) [inline packages/laconia-event/src/DynamoDbStreamRecord.js:19]: Where this come from?

File: packages/laconia-event/src/DynamoDbStreamRecord.js [hunk 3]
```diff
@@ -1,18 +1,36 @@
-const AWS = require("aws-sdk");
+require("aws-sdk/clients/dynamodb");
```
Comment (hugosenari) [inline packages/laconia-event/src/DynamoDbStreamRecord.js:1]: Why this is better?
It's more about what I can learn from your PR because a never see this before (sorry) :-)

I think this could be better for tree shaking but is ugly (AWS-SDK that is, not your suggestion).

**Comment by hugosenari** [comment]
My absence of knowledge in this context prevents me from give any opinion about if your proposal are good or bad for Laconia. 
So I only left some notes about code.

Thank you for your contribution, maybe @ceilfors, @sakthivel-tw and other could help you more here. :-)

File: packages/laconia-event/src/DynamoDbStreamRecord.js (hunk 1 above)
Comment (Ankcorn) [reply packages/laconia-event/src/DynamoDbStreamRecord.js:6]: Perfect will implement that

File: packages/laconia-event/src/DynamoDbStreamRecord.js (hunk 2 above)
Comment (Ankcorn) [reply packages/laconia-event/src/DynamoDbStreamRecord.js:19]: I added an additional getter as I was just spiking around. I haven't used a getter in is before ( cool stuff!)

Looking at it now the name doesn't seem so thoughtful. Maybe just record would be more appropriate

File: packages/laconia-event/src/DynamoDbStreamRecord.js (hunk 3 above)
Comment (Ankcorn) [reply packages/laconia-event/src/DynamoDbStreamRecord.js:1]: Agreed it's unsightly!!! It's based on this article https://theburningmonk.com/2019/03/just-how-expensive-is-the-full-aws-sdk/

**Comment by Ankcorn** [comment]
@hugosenari Thanks for the review 🤗 feel like I am in the same boat with the lack of context! :)

I wanted to spike this out to get a feel for how to implement it but also to start to look at what the interface here should look like. This here would cause the least changes for me at work to implement which is why I suggested it ( how selfish 😝)

I'm not sure it's the most pleasing api to use, also from a framework perspective could have some downsides. If people have suggestions then I am happy to spike them also so we can test and try them :)

There are other databases outside the was ecosystem that let you receive a stream of  events. If laconia intends to be more platform agnostic then would it make sense to choose something here that could accommodate there event structures too.

**Comment by ceilfors** [comment]
Hey @Ankcorn, apologies for the delay in responding, been mulling over this one in the background. Sharing my design thought on where Laconia stand on DynamoDB adapter: I'm not thinking that we should be platform agnostic, but we should help users to think about how good design may look like (which may make it agnostic, but it's not necessary).

Designing the adapter for DynamoDB is complex because of these variables:
1. DynamoDB Stream Record View: KEYS_ONLY, NEW_IMAGE, OLD_IMAGE, NEW_AND_OLD_IMAGES
2. DynamoDB Stream Event type: modify, insert, remove

There are many opportunities where we can help the users, but we might want to prioritise for only the common use case. I'm afraid I don't know what is the best practice or common use case enough. That said, if you're using DynamoDB stream already, combining your use case and my use case will be good. Can you elaborate your use case more?

---

These are the dump of my thoughts on the opportunities / confusion I had:

- KEYS_ONLY:
  - If DynamoDB stream is configured with this, events that are coming will not have the `image`. From framework level, we can help user pull the image ad-hoc if they need it.
  - If stream
- NEW_IMAGE:
  - What does this mean when stream event type is REMOVE?
- NEW_AND_OLD_IMAGES
  - I presume this is the common use case?

If these are the driving factor, should the framework be smart to derive what the user wants based on these? Super hypothetical API:
- dynamodbAdapter.keysOnly(modify: (keys) => { app logic }, insert: (keys) => { app logic }, insert: (keys) -> { app logic })
- dynamodbAdapter.keysOnlyAndFetch(modify: (keys) => { app logic }, insert: (keys) => { app logic }, insert: (keys) -> { app logic })
- dynamodbAdapter.newImage(modify: (newImage) => {}, insert: (newImage) // Notice how remove is unsupported)

**Comment by Ankcorn** [comment]
Hey @ceilfors, thanks!

My use case is to stream data from one table out into a number of consumers.

We ship changes to a kinesis firehose stream and a 3rd party API.
We upsert new records into another dynamo table based on one of there attributes

A table can only have 1 stream. I don't know if this changes what the stream types are but because of this, we need New and Old images even though the 3rd stream only needs keys only.

> NEW_IMAGE:
    What does this mean when stream event type is REMOVE?

My guess would be the new image would be undefined but you would still see the keys of the image? This could be a bit of a weird edge case to handle. I will put together an experiment to test these scenarios.


> keysOnlyAndFetch

what would this method do above the keysOnly method?

I like your api design. I guess at a pseudo-code level our code would go from one big handler to

```javascript
const laconia = require('');
const dynamo = require('').newAndOldImages;

async function shipChangesTo3rdParty(params) {
    await request('', params);
    return 🚢
}

async function remove(oldImage, { shipChangesTo3rdParty }) {
    const params = { ...oldImage } // some data parsing
    await shipChangesTo3rdParty(params);
}

async function insert(newImage, { shipChangesTo3rdParty }) {
    const params = { ...oldImage } // some data parsing
    await shipChangesTo3rdParty(params);
}

async function modify(newImage, oldImage, { shipChangesTo3rdParty }) {
    const params = { ...oldImage, newImage } // some data parsing
    await shipChangesTo3rdParty(params);
}

module.exports.handler = laconia(dynamo({ insert, modify, remove }).register(() => ({ shipChangesTo3rdParty });
```

This generally feels much better than what we already have. The behaviour that would be hard for us to recreate with this API (although I'm not sure we would want too) is that we currently map over the images using some promise lib that lets us run them in batches. We would probably be better off choosing the batch sizes we wanted at an event level.

**Comment by Ankcorn** [comment]
See this repo to verify some assumptions about stream behaviour https://github.com/Ankcorn/dynamodb-streams-test

File: packages/laconia-event/src/DynamoDbStreamRecord.js (hunk 2 above)
Comment (hugosenari) [reply packages/laconia-event/src/DynamoDbStreamRecord.js:19]: You did not set any value as eventName :)

**Comment by hugosenari** [comment]
@Ankcorn here a simple  table of your data:

  | INSERT | MODIFY | REMOVE
-- | -- | -- | --
KEYS_ONLY | Keys | Keys | Keys
NEW_IMAGE | Keys, NewImage | Keys, NewImage | Keys
OLD_IMAGE | Keys | Keys, OldImage | Keys, OldImage
NEW_AND_OLD_IMAGES | Keys, NewImage | Keys, NewImage, OldImage | Keys, OldImage

And data looks like this:

```yaml
Records:
  - eventName: INSERT|MODIFY|REMOVE
    dynamodb:
      Keys: {} # always exists
      NewImage: {} # sometimes exists, see table
      OldImage: {} # sometimes exists, see table
      StreamViewType: KEYS_ONLY|NEW_IMAGE|OLD_IMAGE|NEW_AND_OLD_IMAGES
```

Thank you for the data, will help-me with other projects ;-)

**Comment by Ankcorn** [comment]
Going to update my readme with that thanks! 👍

**Comment by Ankcorn** [comment]
I assume all the stream types exist to make sure you are not incurring a large bill if the objects you are reading from the stream are big?

Do you think this is the best option for the api? I probably have the time this week to do some midnight coding! 🦉

```javascript
laconia(dynamo({ insert: (keys, new, old) => {}, modify: (keys, new, old) => {}, remove: (keys, new, old) => {} })
```

or


```javascript
laconia(dynamo(event: (keys, type, new, old) => {})
```


I like dynamodbAdapter.keysOnlyAndFetch as an idea but I think you would only use this for certain conditions. I.E. the pk or sk match something you are interested in. 

It seems better to leave fetching the record to a separate dynamodb client so not to overcomplicate things

**Comment by hugosenari** [comment]
Suggestion

```javascript
const handlers = {
  insert: async (laconiaContext) => async (dynamodb, record, event) => {}, 
  remove: async (laconiaContext) => async (dynamodb, record, event) => {},
  modify: async (laconiaContext) => async (dynamodb, record, event) => {}
};

laconia(dynamo(handlers))
```
Expecting dynamodb, record, event to be:
```javascript
const event = {
  Records: [{
    eventID: "0000000000",
    eventName: "INSERT",
    eventVersion: "1.1",
    eventSource: "aws:dynamodb",
    awsRegion: "eu-west-1",
    dynamodb: {
      ApproximateCreationDateTime: 0000000000,
      Keys: { },    // always exists
      NewImage: {}, // sometimes exists, see table
      OldImage: {}, // sometimes exists, see table
      SequenceNumber: "0000000000",
      SizeBytes: 0000000000,
      StreamViewType: "KEYS_ONLY"
    },
    eventSourceARN: "arn:"
}];

const records = event.Records;
const [ record ] = records;
const { dynamodb } = record
```

**Comment by Ankcorn** [comment]
In that example is `record` just the first or do we call the callback for each?

**Comment by hugosenari** [comment]
For each.

Unless that user needs to reduce records :thinking:

**Comment by Ankcorn** [comment]
Oh cool that issue is super useful to know about for now. Will work on just a single record. I like the escaspe hatches in your api suggestion so will go with that

**Comment by hugosenari** [comment]
Talking about laconia/batch (#2), when it knows there more work to be done but not 'enough' time it does 'aws lambda recursion'. What should you do if your stream processor couldn't process all records of event?

And if some record fail, all records failed?

**Comment by ceilfors** [comment]
_Apologies for my slow reply, for some reason GitHub has no longer send a notification to my email! Still figuring out how to fix this_

I'm quite caught up with my current startup, hopefully I can get around into this one soon as this one is quite deep 🤞

**Comment by Ankcorn** [comment]
Hey, no worries.

I have been too unmotivated to do anything here.

Hopefully I find the spark again :D

**Comment by ceilfors** [comment]
Please reopen when you found the spark :)
